README.md

```diff
@@ -1,35 +1,455 @@
+
+- [`Loc`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html),
+  providing an abstraction of _shared memory locations_,
+
+- [`Op`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Op/index.html),
+  providing an interface for _primitive operations_ over multiple shared memory
+  locations, and
+
+- [`Tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Tx/index.html),
+  providing _composable transactions_ over shared memory locations.
+
+The following sections discuss each of the above in turn.
+
+### Creating and manipulating individual shared memory locations
+
+The [`Loc`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html)
+module is essentially compatible with the Stdlib
+[`Atomic`](https://v2.ocaml.org/api/Atomic.html) module, except that a number of
+functions take an optional
+[`backoff`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Backoff/index.html)
+as an argument.
+
+In other words, an application that uses
+[`Atomic`](https://v2.ocaml.org/api/Atomic.html), but then needs to perform
+atomic operations over multiple atomic locations, could theoretically just
+rebind `module Atomic = Loc` and then use the
+[`Op`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Op/index.html)
+and/or
+[`Tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Tx/index.html) APIs
+to perform operations over multiple locations. This should not be done
+just-in-case, however, as, even though **kcas** is efficient, it does naturally
+have higher overhead than the Stdlib
+[`Atomic`](https://v2.ocaml.org/api/Atomic.html).
+
+### Programming with primitive operations
+
+The [`Op`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Op/index.html)
+module is probably most suitable when using **kcas** as a means to design and
+implement new lock-free algorithms.
+
+To program with primitive operations one simply makes a list of CAS operations
+using
+[`make_cas`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Op/index.html#val-make_cas)
+and then attempts them using
+[`atomically`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Op/index.html#val-atomically).
+Typically that needs to be done inside a loop of some kind as such an attempt
+can naturally fail.
+
+Let's first
+[`make`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html#val-make)
+two locations representing stacks:
+
+```ocaml
+# let stack_a = Loc.make [19]
+  and stack_b = Loc.make [76]
+val stack_a : int list Loc.t = <abstr>
+val stack_b : int list Loc.t = <abstr>
+```
+
+Here is a function that can atomically move an element from given `source` stack
+to the given `target` stack:
+
+```ocaml
+# let rec move ?(backoff = Backoff.default)
+               source
+               target =
+    match Loc.get source with
+    | [] -> raise Exit
+    | (elem::rest) as source' ->
+      let target' = Loc.get target in
+      let ops = [
+        Op.make_cas source source' rest;
```

Comment:
The use of `source'` to mean "the old value of `source`" is the opposite to other uses I've seen of `'`, which might be confusing.

Reply:
Hmm... How about `source_loc` and `target_loc` for the parameters?

Reply:
Or perhaps `old_source` and `old_target`, instead.  I guess that is clearer.